File: src/runtime/disco/session.cc
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
#include "./session.h"

#include <tvm/runtime/packed_func.h>
#include <tvm/runtime/registry.h>

#include "./worker.h"

namespace tvm {
namespace runtime {

TVM_REGISTER_OBJECT_TYPE(DRefObj);
TVM_REGISTER_OBJECT_TYPE(CommandQueueObj);
TVM_REGISTER_OBJECT_TYPE(SessionObj);

DRefObj::~DRefObj() { Downcast<Session>(this->session)->DeallocReg(reg_id); }

DRef::DRef(int reg_id, Session session) {
  ObjectPtr<DRefObj> p = make_object<DRefObj>();
  p->reg_id = reg_id;
  p->session = session;
  data_ = std::move(p);
}

TVMRetValue DRefObj::DebugGetFromRemote(int worker_id) {
  return Downcast<Session>(this->session)->DebugGetFromRemote(this->reg_id, worker_id);
}

void CommandQueueObj::SyncWorker(int worker_id) {
  this->BroadcastUnpacked(DiscoAction::kSyncWorker, worker_id);
  TVMArgs args = this->RecvReplyPacked(worker_id);
  ICHECK_EQ(args.size(), 2);
  DiscoAction action = static_cast<DiscoAction>(args[0].operator int());
  int ret_worker_id = args[1];
  ICHECK(action == DiscoAction::kSyncWorker);
  ICHECK_EQ(ret_worker_id, worker_id);
}

int SessionObj::AllocateReg() {
  if (this->free_regs_.empty()) {
    return this->reg_count_++;
  }
  int reg_id = this->free_regs_.back();
  this->free_regs_.pop_back();
  return reg_id;
}

void SessionObj::DeallocReg(int reg_id) {
  cmd_->BroadcastUnpacked(DiscoAction::kKillReg, reg_id);
  this->free_regs_.push_back(reg_id);
}

DRef SessionObj::CallPackedWithPacked(const TVMArgs& args) {
  constexpr uint64_t one = 1;
  constexpr int offset = 4;
  TVMValue* values = const_cast<TVMValue*>(args.values);
  int* type_codes = const_cast<int*>(args.type_codes);
  int num_args = args.num_args;
  int reg_id = AllocateReg();
  uint64_t is_dref = 0;
  CHECK_LE(num_args - offset, 64) << "ValueError: Too many arguments, maximum is 64, but got "
                                  << (num_args - offset);
  TVMArgsSetter setter(values, type_codes);
  for (int i = offset; i < num_args; ++i) {
    TVMValue t_value = values[i];
    int type_code = type_codes[i];
    CHECK(type_code != kTVMDLTensorHandle)
        << "ValueError: Cannot pass DLTensor to Session.CallPacked";
    CHECK(type_code != kTVMModuleHandle) << "ValueError: Cannot pass Module to Session.CallPacked";
    CHECK(type_code != kTVMPackedFuncHandle)
        << "ValueError: Cannot pass PackedFunc to Session.CallPacked";
    CHECK(type_code != kTVMNDArrayHandle)
        << "ValueError: Cannot pass NDArray to Session.CallPacked";
    CHECK(type_code != kTVMObjectRValueRefArg)
        << "ValueError: Cannot pass RValue to Session.CallPacked";
    if (type_code == kTVMObjectHandle) {
      TVMArgValue arg(t_value, type_code);
      if (arg.IsObjectRef<DRef>()) {
        setter(i, arg.AsObjectRef<DRef>()->reg_id);
        is_dref |= one << (i - offset);
      } else if (arg.IsObjectRef<String>()) {
        // NOTE: String is always on heap during `cmd_->BroadcastPacked` because there is no chance
        // its reference counter decreases.
        setter(i, arg.AsObjectRef<String>()->data);
      } else {
        LOG(FATAL) << "ValueError: Cannot pass Object to Session.CallPacked";
      }
    }
  }
  DRef func = args[2];
  setter(0, static_cast<int>(DiscoAction::kCallPacked));
  setter(1, reg_id);
  setter(2, func->reg_id);
  setter(3, is_dref);
  cmd_->BroadcastPacked(TVMArgs(values, type_codes, num_args));
  return DRef(reg_id, GetRef<Session>(this));
}

DRef SessionObj::GetGlobalFunc(const std::string& name) {
  int reg_id = AllocateReg();
  cmd_->BroadcastUnpacked(DiscoAction::kGetGlobalFunc, reg_id, name);
  return DRef(reg_id, GetRef<Session>(this));
}

void SessionObj::CopyFromWorker0(const NDArray& host_array, const DRef& remote_array) {
  cmd_->AppendHostNDArray(host_array);
  cmd_->BroadcastUnpacked(DiscoAction::kCopyFromWorker0, remote_array->reg_id);
}

void SessionObj::CopyToWorker0(const NDArray& host_array, const DRef& remote_array) {
  cmd_->AppendHostNDArray(host_array);
  cmd_->BroadcastUnpacked(DiscoAction::kCopyToWorker0, remote_array->reg_id);
}

void SessionObj::SyncWorker(int worker_id) { this->cmd_->SyncWorker(worker_id); }

TVMRetValue SessionObj::DebugGetFromRemote(int64_t reg_id, int worker_id) {
  return this->cmd_->DebugGetFromRemote(reg_id, worker_id);
}

struct Internal {
  static DRef CallPackedWithPacked(Session sess, const TVMArgs& args) {
    return sess->CallPackedWithPacked(args);
  }
};

TVM_REGISTER_GLOBAL("runtime.disco.SessionThreaded").set_body_typed(Session::ThreadedSession);
TVM_REGISTER_GLOBAL("runtime.disco.DRefSession").set_body_typed([](DRef obj) {
  return obj->session;
});
TVM_REGISTER_GLOBAL("runtime.disco.DRefDebugGetFromRemote")
    .set_body_method<DRef>(&DRefObj::DebugGetFromRemote);
TVM_REGISTER_GLOBAL("runtime.disco.SessionGetGlobalFunc")
    .set_body_method<Session>(&SessionObj::GetGlobalFunc);
TVM_REGISTER_GLOBAL("runtime.disco.SessionCopyFromWorker0")
    .set_body_method<Session>(&SessionObj::CopyFromWorker0);
TVM_REGISTER_GLOBAL("runtime.disco.SessionCopyToWorker0")
    .set_body_method<Session>(&SessionObj::CopyToWorker0);
TVM_REGISTER_GLOBAL("runtime.disco.SessionSyncWorker")
    .set_body_method<Session>(&SessionObj::SyncWorker);
TVM_REGISTER_GLOBAL("runtime.disco.SessionCallPacked").set_body([](TVMArgs args, TVMRetValue* rv) {
  Session self = args[0];
  *rv = Internal::CallPackedWithPacked(
      self, TVMArgs(args.values + 1, args.type_codes + 1, args.num_args - 1));
});

}  // namespace runtime
}  // namespace tvm

File: python/tvm/runtime/disco/session.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.
"""This module defines a Session in Disco. Session is the primary interface that users interact
with the distributed runtime.
"""
from typing import Any, Callable, Optional, Sequence

from ..._ffi import register_object
from ..._ffi.runtime_ctypes import Device
from ..ndarray import NDArray
from ..object import Object
from . import _ffi_api


@register_object("runtime.disco.DRef")
class DRef(Object):
    """An object that exists on all workers. The controller process assigns a unique "register id"
    to each object, and the worker process uses this id to refer to the object residing on itself.
    """

    @property
    def session(self) -> "Session":
        """Get the session that this DRef belongs to."""
        return _ffi_api.DRefSession(self)  # type: ignore # pylint: disable=no-member

    def debug_get_from_remote(self, worker_id: int) -> Any:
        """Get the value of a DRef from a remote worker.

        Parameters
        ----------
        worker_id : int
            The id of the worker to be fetched from.

        Returns
        -------
        value : object
            The value of the register.
        """
        return _ffi_api.DRefDebugGetFromRemote(self, worker_id)  # type: ignore # pylint: disable=no-member


class DPackedFunc(DRef):
    """A PackedFunc in a Disco session."""

    def __init__(self, dref: DRef) -> None:
        self.handle = dref.handle
        dref.handle = None

    def __call__(self, *args) -> DRef:
        return self.session.call_packed(self, *args)


class DModule(DRef):
    """A Module in a Disco session."""

    def __init__(self, dref: DRef) -> None:
        self.handle = dref.handle
        del dref.handle

    def __getitem__(self, name: str) -> DPackedFunc:
        func = self.session._get_cached_method("runtime.ModuleGetFunction")
        return DPackedFunc(func(self, name, False))


@register_object("runtime.disco.Session")
class Session(Object):
    """A Disco interactive session. It allows users to interact with the Disco command queue with
    various PackedFunc calling convention."""

    def init_ccl(self, api: str, *args):
        """Initialize the underlying communication collective library.

        Parameters
        ----------
        api : str
            The name of the communication collective library. Currently supported libraries are:
            - nccl
            - rccl
            - mpi
        *args : various types
            The arguments to be passed to the initialization function of the communication
        """
        assert api in ("nccl", "rccl"), f"Unsupported CCL backend: {api}"
        func = self.get_global_func(f"runtime.disco.{api}.init")
        func(*args)

    @staticmethod
    def threaded_session(num_workers: int) -> "Session":
        """Create a threaded session."""
        return _ffi_api.SessionThreaded(num_workers)  # type: ignore # pylint: disable=no-member

    def _get_cached_method(self, name: str) -> Callable:
        if not hasattr(self, "_cache"):
            cache = self._cache = {}  # pylint: disable=attribute-defined-outside-init
        else:
            cache = self._cache
        if name not in cache:
            func = cache[name] = self.get_global_func(name)
        else:
            func = cache[name]
        return func

    def empty(
        self,
        shape: Sequence[int],
        dtype: str,
        device: Optional[Device] = None,
    ) -> DRef:
        """Create an empty NDArray on all workers.

        Parameters
        ----------
        shape : tuple of int
            The shape of the NDArray.
        dtype : str
            The data type of the NDArray.
        device : Optional[Device] = None
            The device of the NDArray.

        Returns
        -------
        array : DRef
            The created NDArray.
        """
        if device is None:
            device = Device(device_type=0, device_id=0)
        func = self._get_cached_method("runtime.disco.empty")
        return func(*shape, dtype, device)

    def get_global_func(self, name: str) -> DRef:
        """Get a global function on workers.

        Parameters
        ----------
        name : str
            The name of the global function.

        Returns
        -------
        func : DRef
            The global packed function
        """
        return DPackedFunc(_ffi_api.SessionGetGlobalFunc(self, name))  # type: ignore # pylint: disable=no-member

    def call_packed(self, func: DRef, *args) -> DRef:
        """Call a PackedFunc on workers providing variadic arguments.

        Parameters
        ----------
        func : PackedFunc
            The function to be called.
        *args : various types
            In the variadic arguments, the supported types include:
            - integers and floating point numbers;
            - DLDataType;
            - DLDevice;
            - str (std::string in C++);
            - DRef.

        Returns
        -------
        return_value : various types
            The return value of the function call.

        Notes
        -----
        Examples of unsupported types:
        - NDArray, DLTensor;
        - TVM Objects, including PackedFunc and Module.
        """
        return _ffi_api.SessionCallPacked(self, 0, 0, func, 0, *args)  # type: ignore # pylint: disable=no-member

    def sync_worker(self, worker_id: int = 0) -> None:
        """Synchronize the controller with a worker, and it will wait until the worker finishes
        executing this instruction.

        Parameters
        ----------
        worker_id : int
            The id of the worker to be synced with.

        Notes
        -----
        This function is usually used for worker-0, because it is the only worker that is
        assumed to collocate with the controller. Syncing with other workers may not be supported
        and should only be used for debugging purposes.
        """
        return _ffi_api.SessionSyncWorker(self, worker_id)  # type: ignore # pylint: disable=no-member

    def copy_from_worker_0(self, host_array: NDArray, remote_array: DRef) -> None:
        """Copy the controller-side NDArray to worker-0.

        Parameters
        ----------
        host_array : numpy.ndarray
            The array to be copied to worker-0.
        remote_array : NDArray
            The NDArray on worker-0.
        """
        return _ffi_api.SessionCopyFromWorker0(self, host_array, remote_array)  # type: ignore # pylint: disable=no-member

    def copy_to_worker_0(self, host_array: NDArray, remote_array: DRef) -> None:
        """Copy an NDArray from worker-0 to the controller-side NDArray.

        Parameters
        ----------
        host_array : numpy.ndarray
            The array to be copied from worker-0.
        remote_array : NDArray
            The NDArray on worker-0.
        """
        return _ffi_api.SessionCopyToWorker0(self, host_array, remote_array)  # type: ignore # pylint: disable=no-member

    def load_vm_module(
        self,
        path: str,
        device: Optional[Device] = None,
    ) -> DModule:
        """Load a VM module from a file.

        Parameters
        ----------
        path : str
            The path to the VM module file.
        device : Optional[Device] = None
            The device to load the VM module to. Default to the default device of each worker.

        Returns
        -------
        module : DModule
            The loaded VM module.
        """
        if device is None:
            device = Device(device_type=0, device_id=0)
        func = self._get_cached_method("runtime.disco.load_vm_module")
        return DModule(func(path, device))
